### PR TITLE
Use sanitize-prow-jobs to control jobs' cluster

### PR DESCRIPTION
**Pull request by hongkailiu**
/cc @alvaroaleman @petr-muller 

Tested with

```bash
sanitize-prow-jobs --prow-jobs-dir=./ci-operator/jobs
```

/hold

requires https://github.com/openshift/release/pull/8209/files

### CODE REVIEW

File: cmd/sanitize-prow-jobs/main.go [hunk 1]
```diff
@@ -15,27 +17,66 @@ import (
...
 }
 
+type Config struct {
+	Clusters struct {
+		Default      string `json:"default"`
+		Distribution map[string]struct {
+			Jobs   []string `json:"jobs"`
+			Files  []string `json:"files"`
+			Groups []string `json:"groups"`
+		} `json:"distribution"`
```
Comment (alvaroaleman) [inline cmd/sanitize-prow-jobs/main.go:34]: What about `assignment`? The distribution is more the result of the assignment we do here

File: cmd/sanitize-prow-jobs/main.go [hunk 2]
```diff
@@ -15,27 +17,66 @@ import (
...
+		Groups map[string][]string       `json:"groups"`
+		REMap  map[string]*regexp.Regexp `json:"-"`
```
Comment (alvaroaleman) [inline cmd/sanitize-prow-jobs/main.go:36]: This is the REMap for the files, right? Why is it not alongside of files but a level higher? Also please name `FileREMap`, otherwise we have to rename it should we ever need another `REMap`

File: cmd/sanitize-prow-jobs/main.go [hunk 3]
```diff
@@ -97,28 +138,64 @@ func determinizeJobs(prowJobConfigDir string) error {
...
-		jb.Cluster = defaultCluster
+func getCluster(name string, path string, config *Config) string {
+	distribution := config.Clusters.Distribution
+	clusters := make([]string, 0, len(distribution))
+	for cluster := range distribution {
+		clusters = append(clusters, cluster)
 	}
+	sort.Strings(clusters)
 
+	//jobs override files which override groups
```
Comment (alvaroaleman) [inline cmd/sanitize-prow-jobs/main.go:169]: Please add this comment to the type definition

File: cmd/sanitize-prow-jobs/main.go [hunk 4]
```diff
@@ -97,28 +138,64 @@ func determinizeJobs(prowJobConfigDir string) error {
...
 }
 
-func defaultJobBase(jb *prowconfig.JobBase) {
+func defaultJobBase(jb *prowconfig.JobBase, path string, config *Config) {
+	jb.Cluster = getCluster(jb.Name, path, config)
+}
 
-	if jb.Cluster == "" || jb.Cluster == "default" {
-		jb.Cluster = defaultCluster
+func getCluster(name string, path string, config *Config) string {
```
Comment (alvaroaleman) [inline cmd/sanitize-prow-jobs/main.go:161]: Please rename `getCluster` -> `getClusterForJob` and `name` -> `jobName`.

Would it maybe make sense to make `getCluster` a method of `config` to have them closer together and signal the tight coupling?

File: test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+---
```
Comment (alvaroaleman) [inline test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml:1]: Why is this needed?

File: cmd/sanitize-prow-jobs/main_test.go [hunk 6]
```diff
@@ -19,7 +20,20 @@ func TestDefaultJobConfig(t *testing.T) {
 		Periodics: []config.Periodic{{}, {}, {JobBase: config.JobBase{Cluster: "default"}}},
 	}
 
-	defaultJobConfig(jc)
+	config := &Config{Clusters: struct {
+		Default      string `json:"default"`
```
Comment (alvaroaleman) [inline cmd/sanitize-prow-jobs/main_test.go:24]: I don't think this needs json tags

File: cmd/sanitize-prow-jobs/main.go [hunk 7]
```diff
@@ -97,28 +138,64 @@ func determinizeJobs(prowJobConfigDir string) error {
...
+			if name == job {
+				return cluster
+			}
+		}
+	}
+	for _, cluster := range clusters {
+		for _, file := range distribution[cluster].Files {
+			re, ok := config.Clusters.REMap[file]
+			if !ok {
+				continue
```
Comment (alvaroaleman) [inline cmd/sanitize-prow-jobs/main.go:181]: I think this should never happen and hence this is an error or not?

File: cmd/sanitize-prow-jobs/main.go [hunk 8]
```diff
@@ -15,27 +17,66 @@ import (
...
-func determinizeJobs(prowJobConfigDir string) error {
+func determinizeJobs(prowJobConfigDir, configPath string) error {
 	errChan := make(chan error)
-
 	var errs []error
+
+	config := &Config{}
+	data, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		errs = append(errs, err)
```
Comment (alvaroaleman) [inline cmd/sanitize-prow-jobs/main.go:57]: I think we should return the error here and in the unmarshaling, as all forthcoming errors may be a result of erroring here. Also please wrap them so they are easier to understand and can be grepped for in the codebase

File: cmd/sanitize-prow-jobs/main.go (hunk 1 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: cmd/sanitize-prow-jobs/main.go (hunk 2 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Renamed.

> Why is it not alongside of files but a level higher?
That was my first try too. I think the current structure is simpler.

File: cmd/sanitize-prow-jobs/main.go (hunk 3 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: cmd/sanitize-prow-jobs/main.go (hunk 4 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml (hunk 5 above)
Comment (hongkailiu) [reply test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml]: Removed.

File: cmd/sanitize-prow-jobs/main.go [hunk 9]
```diff
@@ -6,36 +6,120 @@ import (
...
-
 type options struct {
 	prowJobConfigDir string
+	configPath       string
 
 	help bool
 }
 
+type Config struct {
+	Clusters struct {
```
Comment (stevekuznetsov) [inline cmd/sanitize-prow-jobs/main.go:29]: Any reason that these are anonymous types?

File: cmd/sanitize-prow-jobs/main.go [hunk 10]
```diff
@@ -6,36 +6,120 @@ import (
...
+		Default    string `json:"default"`
+		Assignment map[string]struct {
+			// Jobs override both files and groups
+			Jobs []string `json:"jobs"`
+			// Files override files
```
Comment (stevekuznetsov) [inline cmd/sanitize-prow-jobs/main.go:34]: ?

File: cmd/sanitize-prow-jobs/main.go [hunk 11]
```diff
@@ -6,36 +6,120 @@ import (
...
 type options struct {
 	prowJobConfigDir string
+	configPath       string
 
 	help bool
 }
 
+type Config struct {
+	Clusters struct {
+		Default    string `json:"default"`
```
Comment (stevekuznetsov) [inline cmd/sanitize-prow-jobs/main.go:30]: Public-facing config fields should have GoDoc

File: cmd/sanitize-prow-jobs/main.go [hunk 12]
```diff
@@ -6,36 +6,120 @@ import (
...
+	} `json:"clusters"`
+}
+
+func (config *Config) getClusterForJob(jobName string, path string) string {
+	distribution := config.Clusters.Assignment
+	clusters := make([]string, 0, len(distribution))
+	for cluster := range distribution {
+		clusters = append(clusters, cluster)
+	}
+	sort.Strings(clusters)
```
Comment (stevekuznetsov) [inline cmd/sanitize-prow-jobs/main.go:49]: Why?

File: cmd/sanitize-prow-jobs/main.go [hunk 13]
```diff
@@ -6,36 +6,120 @@ import (
...
+
+	//jobs override files which override groups
+	for _, cluster := range clusters {
+		for _, job := range distribution[cluster].Jobs {
+			if jobName == job {
+				return cluster
+			}
+		}
+	}
+	for _, cluster := range clusters {
```
Comment (stevekuznetsov) [inline cmd/sanitize-prow-jobs/main.go:59]: This smells like a bad configuration structure. It should not be this hard to determine the cluster for a job. Why is it even possible to list a job in multiple lists/groups/regex-matches?

Why not:

```yaml
groups:
- name: foo
  jobs:
  - bar
  cluster: zing
raw:
- jobs:
  - whoa
  cluster: there
files:
- paths:
  - ^whatever.yaml$
  cluster: boogie
```

Comment (stevekuznetsov) [reply cmd/sanitize-prow-jobs/main.go:59]: Or, validate on config load that a job is not in more than one bucket and take it's presence in a bucket as authoritative -- that way if you find it under `cluster.files` or `cluster.groups` you know it belongs in that cluster no matter what.

Comment (stevekuznetsov) [reply cmd/sanitize-prow-jobs/main.go:59]: Said another way, we are building a lot of complexity here to allow `cluster1.groups` to override `cluster2.jobs` which seems silly

File: cmd/sanitize-prow-jobs/main.go [hunk 14]
```diff
@@ -6,36 +6,120 @@ import (
...
 	var errs []error
+	fileREMap := map[string]*regexp.Regexp{}
+	for _, v := range config.Clusters.Assignment {
+		for _, file := range v.Files {
+			if _, ok := fileREMap[file]; ok {
+				continue
+			}
+			re, err := regexp.Compile(file)
+			if err != nil {
+				errs = append(errs, err)
```
Comment (stevekuznetsov) [inline cmd/sanitize-prow-jobs/main.go:115]: without putting the cluster name, assignmment, and file name into this error it will be impossible to debug

File: cmd/sanitize-prow-jobs/main.go [hunk 15]
```diff
@@ -97,28 +181,24 @@ func determinizeJobs(prowJobConfigDir string) error {
...
 	}
 }
 
-func defaultJobBase(jb *prowconfig.JobBase) {
-
-	if jb.Cluster == "" || jb.Cluster == "default" {
-		jb.Cluster = defaultCluster
-	}
-
+func defaultJobBase(jb *prowconfig.JobBase, path string, config *Config) {
```
Comment (stevekuznetsov) [inline cmd/sanitize-prow-jobs/main.go:200]: Why do we want a stateful function with a side-effect?

File: cmd/sanitize-prow-jobs/main_test.go [hunk 16]
```diff
@@ -19,7 +20,20 @@ func TestDefaultJobConfig(t *testing.T) {
 		Periodics: []config.Periodic{{}, {}, {JobBase: config.JobBase{Cluster: "default"}}},
 	}
 
-	defaultJobConfig(jc)
+	config := &Config{Clusters: struct {
+		Default    string `json:"default"`
```
Comment (stevekuznetsov) [inline cmd/sanitize-prow-jobs/main_test.go:24]: See and *this* is why anonymous structs have a very limited use-case...

File: test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml [hunk 17]
```diff
@@ -0,0 +1,2 @@
+clusters:
+  default: api.ci
```
Comment (stevekuznetsov) [inline test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml:2]: We're going to need something *way* more complicated to test this :)

File: cmd/sanitize-prow-jobs/main_test.go (hunk 6 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main_test.go]: Done.

File: cmd/sanitize-prow-jobs/main.go (hunk 7 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: cmd/sanitize-prow-jobs/main.go (hunk 8 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: cmd/sanitize-prow-jobs/main.go (hunk 10 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: cmd/sanitize-prow-jobs/main.go (hunk 12 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: No need to do this with the new structure.

File: cmd/sanitize-prow-jobs/main.go (hunk 13 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: cmd/sanitize-prow-jobs/main.go (hunk 9 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: cmd/sanitize-prow-jobs/main.go (hunk 11 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: cmd/sanitize-prow-jobs/main.go (hunk 14 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: cmd/sanitize-prow-jobs/main_test.go (hunk 16 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main_test.go]: Done.

File: cmd/sanitize-prow-jobs/main.go (hunk 15 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Removed.

File: test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml (hunk 17 above)
Comment (hongkailiu) [reply test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml:2]: (Tested with the config defined https://github.com/openshift/release/pull/8209/files)

Will add unit tests when the core code goes review.

Comment (stevekuznetsov) [reply test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml:2]: Unit and integration tests should exist before review, not after -- they are just as important as code.

File: cmd/sanitize-prow-jobs/main.go [hunk 18]
```diff
@@ -6,36 +6,134 @@ import (
...
+	for _, g := range config.Clusters.Raw {
+		for _, job := range g.Jobs {
+			if jobName == job {
+				return g.Cluster
+			}
+		}
+	}
+	for _, file := range config.Clusters.Files {
+		for _, p := range file.Paths {
+			if config.Clusters.FileREMap[p].MatchString(path) {
```
Comment (stevekuznetsov) [inline cmd/sanitize-prow-jobs/main.go:78]: co-locate the regex and the struct that created it

```
if file.regex.MatchString()
```

**Comment by stevekuznetsov** [comment]
Do we even need regex file matching? Can't we achieve the same thing with regex presubmit name matching?

File: cmd/sanitize-prow-jobs/main.go [hunk 19]
```diff
@@ -6,36 +6,134 @@ import (
...
 }
 
+// Config defines the structure of configuration file for this tool
+type Config struct {
+	Clusters Clusters
+}
+
+// Clusters defines the cluster parameter for each Prow job, i.e., where it runs
+type Clusters struct {
+	// the cluster context name if no other condition matches
```
[the remark this answers is not in the document]
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go:34]: Done.

File: cmd/sanitize-prow-jobs/main.go [hunk 20]
```diff
@@ -6,36 +6,134 @@ import (
...
+// Group defines a name for a group of jobs and where they run
+type Group struct {
+	// the name of the group
+	Name string
+	// the job names of the group
+	UnnamedGroup `json:",inline"`
+}
+
+// UnnamedGroup defines where a group of jobs run
+type UnnamedGroup struct {
```
[the remark this answers is not in the document]
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: I chose the one with name to remind us of what the group is about/from.

File: cmd/sanitize-prow-jobs/main.go (hunk 2 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: cmd/sanitize-prow-jobs/main.go (hunk 18 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: cmd/sanitize-prow-jobs/main.go [hunk 21]
```diff
@@ -6,36 +6,134 @@ import (
...
 type options struct {
 	prowJobConfigDir string
+	configPath       string
 
 	help bool
 }
 
+// Config defines the structure of configuration file for this tool
+type Config struct {
+	Clusters Clusters
```
[the remark this answers is not in the document]
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go:29]: I think it is from https://github.com/openshift/release/pull/8209#pullrequestreview-389839352

**Comment by hongkailiu** [comment]
> Do we even need regex file matching? Can't we achieve the same thing with regex presubmit name matching?

Comparing job names and job file paths, the latter is more regular.
I feel it is easier to express something like, keep the periodics of the release repo on api.ci.

File: test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml (hunk 17 above)
Comment (hongkailiu) [reply test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml:2]: I am not saying it is not as important.
Just feel that it is more efficient after the core code is more stable.
I meant to do it with the current PR, not following up.

Comment (hongkailiu) [reply test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml:2]: Done.

File: cmd/sanitize-prow-jobs/main.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply cmd/sanitize-prow-jobs/main.go:29]: Maybe it looked different before, but there is no utility anymore

[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply cmd/sanitize-prow-jobs/main.go:29]: This is outstanding @hongkailiu

File: cmd/sanitize-prow-jobs/main.go [hunk 22]
```diff
@@ -6,36 +6,126 @@ import (
...
+}
+
+// Group is where a group of jobs run
+type Group struct {
+	// name of the group
+	Name string
+	// the job names
+	Jobs []string
+	// the cluster where the jobs run
+	Cluster string
```
Comment (alvaroaleman) [inline cmd/sanitize-prow-jobs/main.go:50]: Why does this field exist? The group should just have a name and a list of jobs. The `Cluster` then specifies the `group(s)`.

Same comment for the `File` struct below.

File: cmd/sanitize-prow-jobs/main.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go:29]: Missed this. On it now.

[the remark this answers is not in the document]
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: cmd/sanitize-prow-jobs/main.go (hunk 22 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go:45]: So you want to define cluster on groups instead of each group?

File: cmd/sanitize-prow-jobs/main.go [hunk 23]
```diff
@@ -6,36 +6,121 @@ import (
...
 	prowJobConfigDir string
+	configPath       string
 
 	help bool
 }
 
+// Config is the configuration file of this tools, which defines the cluster parameter for each Prow job, i.e., where it runs
+type Config struct {
+	// the cluster context name if no other condition matches
+	Default string
```
Comment (stevekuznetsov) [inline cmd/sanitize-prow-jobs/main.go:30]: Did we lose the JSON tags?

File: cmd/sanitize-prow-jobs/main.go (hunk 22 above)
Comment (stevekuznetsov) [reply cmd/sanitize-prow-jobs/main.go:45]:
```yaml
default: "foo"
groups:
  "bar":
    paths:
    - etc
    - what
    jobs:
    - thing
    - other
```

This seems like the simplest approach. no?

File: cmd/sanitize-prow-jobs/main.go (hunk 23 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go:30]: "sigs.k8s.io/yaml"

This pkg seems working without json tags for those fields.
(tested with config in https://github.com/openshift/release/pull/8209)

File: cmd/sanitize-prow-jobs/main.go (hunk 22 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go:45]: Will do this tonight.

Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go:45]: Asked my question
https://coreos.slack.com/archives/GB7NB0CUC/p1587421714096000

File: cmd/sanitize-prow-jobs/main.go (hunk 23 above)
Comment (stevekuznetsov) [reply cmd/sanitize-prow-jobs/main.go:30]: Sure, there are defaults when your field names are the same, but please follow the convention we use literally everywhere else -- explicitly set them.

Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: cmd/sanitize-prow-jobs/main.go (hunk 22 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: cmd/sanitize-prow-jobs/main.go [hunk 24]
```diff
@@ -132,7 +199,14 @@ func main() {
 	}
 
 	if len(opt.prowJobConfigDir) > 0 {
-		if err := determinizeJobs(opt.prowJobConfigDir); err != nil {
+		if len(opt.configPath) == 0 {
```
Comment (stevekuznetsov) [inline cmd/sanitize-prow-jobs/main.go:202]: We should not do this -- always expect a user to give the flag and error if not.

Comment (stevekuznetsov) [reply cmd/sanitize-prow-jobs/main.go:202]: For seamless roll-out, we need a new PR first that adds the flag but does not validate that it is set. Then, we add a PR to add the flag to the job. Then, we land this PR, where we enforce it is set and use it

File: test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml [hunk 25]
```diff
@@ -0,0 +1,16 @@
+default: api.ci
+nonKubernetes: api.ci
+groups:
```
Comment (stevekuznetsov) [inline test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml:3]: you dont need the map for these files

File: test/repo-init-integration/expected/ci-operator/jobs/org/other/org-other-nonstandard-presubmits.yaml [hunk 26]
```diff
@@ -4,7 +4,7 @@ presubmits:
     always_run: true
     branches:
     - nonstandard
-    cluster: api.ci
+    cluster: ci/api-build01-ci-devcluster-openshift-com:6443
```
Comment (stevekuznetsov) [inline test/repo-init-integration/expected/ci-operator/jobs/org/other/org-other-nonstandard-presubmits.yaml:7]: Why did a `ci-operator` test change?

Comment (hongkailiu) [reply test/repo-init-integration/expected/ci-operator/jobs/org/other/org-other-nonstandard-presubmits.yaml:7]: Because we are testing the config in the integration test.
The config tells to run that job in build01.

https://github.com/openshift/ci-tools/blob/0fead13f0eb9f78c23e3fff1c7042b9d345421c4/test/repo-init-integration/input/core-services/sanitize-prow-jobs/_config.yaml#L7-L10

File: test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml (hunk 25 above)
Comment (hongkailiu) [reply test/repo-init-integration/expected/core-services/sanitize-prow-jobs/_config.yaml:3]: Removed.

File: cmd/sanitize-prow-jobs/main.go (hunk 24 above)
Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: On it now.

Comment (hongkailiu) [reply cmd/sanitize-prow-jobs/main.go]: Done.

File: cmd/autoconfigbrancher/main.go [hunk 27]
```diff
@@ -199,7 +199,7 @@ func main() {
 	run(cmd, args...)
```
Comment (hongkailiu) [inline cmd/autoconfigbrancher/main.go:199]: Will revert https://github.com/openshift/ci-tools/pull/659/files
when everything is working with sanitize tool
